# Homework 5: Lexicons and Distributional Semantics

This is due on **Friday, 11/10 (11pm)** 

## How to do this problem set

Most of these questions require writing Python code and computing results, and the rest of them have textual answers.  Write all the textual answers in this document, show the output of your experiment in this document, and implement the functions in the `python` files. 

Submit a PDF of thie .ipynb to Gradescope, and the .ipynb and all python files to Moodle.

The assignment has two parts:
 * In the first part, you will experiment with Turney's method to find word polarities in a twitter dataset, given some positive and negative seed words.
 * In the second part, you will experiment with distributional and vector semantics.

Your Name: Krishna Prasad Sankaranarayanan

**List collaborators, and how you collaborated, here:** (see our [grading and policies page](http://people.cs.umass.edu/~brenocon/inlp2017/grading.html) for details on our collaboration policy).

* _name 1_ 

## Part 1: Lexicon semantics


Recall that PMI of a pair of words, is defined as:

$$PMI(x, y) = log\frac{ P(x, y) }{ P(x)P(y)}$$

The Turney mehod defines a word's polarity as:

$$Polarity(word) = PMI(word, positive\_word)−PMI(word, negative\_word)$$

where the joint probability $P(w, v)$ or, more specifically, $P(w\ NEAR\ v)$ is the probability of both being "near" each other.  We'll work with tweets, so it means: if you choose a tweet at random, what's the chance it contains both `w` and `v`?

(If you look at the Turney method as explained in the SLP3 chapter, the "hits" function is a count of web pages that contain at least one instance of the two words occurring near each other.)

The positive_word and negative_word terms are initially constructed by hand. For example: we might start with single positive word ('excellent') and a single negative word ('bad'). We can also have list of positive words ('excellent', 'perfect', 'love', ....) and list of negative words ('bad', 'hate', 'filthy',....)

If we're using a seed list of multiple terms, just combine them into a single symbol, e.g. all the positive seed words get rewritten to POS_WORD (and similarly for NEG_WORD).  This $P(w, POS\_WORD)$ effectively means the co-ocurrence of $w$ with any of the terms in the list.

For this assignment, we will use twitter dataset which has 349378 tweets. These tweets are in the file named `tweets.txt`. These are the tweets of one day and filtered such that each tweet contains at least one of the seed words we've selected.

## Question 1 (15 points)

The file `tweets.txt` contains around 349,378 tweets with one tweet per line.  It is a random sample of public tweets, which we tokenized with [twokenize.py's tokenizeRawTweetText()](https://github.com/myleott/ark-twokenize-py/blob/master/twokenize.py)). The text you see has a space between each token so you can just use `.split()` if you want.  We also filtered tweets to ones that included at least one term from one of these seed lists:
* Positive seed list: ["good", "nice", "love", "excellent", "fortunate", "correct", "superior"]
* Negative seed list: ["bad", "nasty", "poor", "hate", "unfortunate", "wrong", "inferior"]

Each tweet contains at least one positive or negative seed word. Take a look at the file (e.g. `less' and `grep'). Implement the Turney's method to calculate polarity scores of all words.

Some things to keep in mind:
* Ignore the seed words (i.e. don't calculate the polarity of the seed words).
* You may want to ignore the polarity of words beignning with `@` or `#`. 

We recommend that you write code in a python file, but it's up to you.

QUESTION: You'll have to do something to handle zeros in the PMI equation. Please explain your and justify your decision about this.

In Mutual Information, zeroes can occur in the computation of positive and negative seed list. So smoothing is done with value as 0.01 like in naive bayes.

## Question 2 (5 points)

Print the top 50 most-positive words (i.e. inferred positive words) and the 50 most-negative words.

Many of the words won't make sense.  Comment on at least two that do make sense, and at least two that don't.  Why do you think these are happening with this dataset and method?

In [20]:
%run turney.py
print_top50()

Top - 50 Positive Words: 


1 . teyana => 10.6120964668
2 . von => 10.4478644941
3 . keef => 10.4455735686
4 . herban => 10.4444261348
5 . 😂😂😂😍 => 10.4305532403
6 . compatibility => 9.33079945215
7 . calculated => 9.3131629328
8 . meera => 9.05009737982
9 . doghouse => 9.04079541981
10 . doggo… => 9.04079541981
11 . pupared => 9.04079541981
12 . paredes => 9.01715477462
13 . colores => 9.01235883238
14 . arguing\ => 8.98308987803
15 . rp2 => 8.92184937926
16 . ♡ => 8.87883430753
17 . shrek => 8.85097890506
18 . hosting => 8.69871898506
19 . 😍❤ => 8.60209882321
20 . bit… => 8.58750108935
21 . only🍀 => 8.48703832979
22 . memes… => 8.41105884075
23 . viola => 8.2998435246
24 . 💘 => 8.2800428579
25 . *7 => 8.26999353204
26 . lik… => 8.25984219119
27 . 😍they => 8.22875482613
28 . 😬💯 => 8.22875482613
29 . 😂🕺🏽 => 8.22875482613
30 . this😊 => 8.2181738365
31 . 🙄🙄🤷🏽‍♂️ => 8.2181738365
32 . treated… => 8.1966699437
33 . craze => 8.15222312129
34 . 🌴 => 8.14079573149
35 . atlantic => 8.12923624547

Names of people and places as expected get the top polarity in the twitter tweets datasert. Examples are Herban, Teyana. Symbols and 

Emoticon occur on top of the heap since they express emotions and therefore happy emoticons occur with happy positive words. 

In the negative wordslist, terms like crimes,vandalism,anti-muslim occur as negative wordd as expected. On the other hand words like safest and hearts  ar negative becasue of the irregularities inthe twitter dataset with resepct to these topics. Also, multiple tweets occur again and again making the data un curated and therefore affecting the polarity negatively.

## Question 3 (5 points)

Now filter out all the words which have total count < 500, and then print top 50 polarity words and bottom 50 polarity words. 

Choose some of the words from both the sets of 50 words you got above which accoording to you make sense. Again please note, you will find many words which don't make sense. Do you think these results are better than the results you got in Question-1? Explain why.

In [22]:
%run turney.py
print_top50withfilter()

Top - 50 POSITIVIE WORDS:

1 . teyana => 10.6120964668
2 . von => 10.4478644941
3 . keef => 10.4455735686
4 . herban => 10.4444261348
5 . 😂😂😂😍 => 10.4305532403
6 . decay => 4.73846330389
7 . ol => 4.32302479215
8 . taylor => 4.06853339313
9 . pregnancy => 3.75511496517
10 . hood => 3.68472300776
11 . birthday => 3.15133512578
12 . ❤️ => 2.78592672669
13 . 💕 => 2.7534945076
14 . ❤ => 2.6493937563
15 . 😊 => 2.28039673113
16 . hilarious => 2.24080355143
17 . morning => 1.92304674544
18 . d => 1.91532342553
19 . fall => 1.83391953037
20 . thank => 1.70725280998
21 . falling => 1.70125934775
22 . happy => 1.68318373846
23 . sweet => 1.595075618
24 . march => 1.52815929992
25 . beautiful => 1.51526424663
26 . thanks => 1.496764906
27 . nadine => 1.48168016343
28 . follow => 1.42461101173
29 . amazing => 1.42297261604
30 . 😍 => 1.33591430478
31 . proud => 1.33092090515
32 . loved => 1.32996373815
33 . soon => 1.31079919419
34 . feels => 1.18121983239
35 . forever => 1.18047972125
36 . sounds 

It definitely is a big positive for the performance since it accounts for words which coud have occured not even once for a negative context and vice versa. This makes the polarity score giving strong priority for the occurence class making it a better prediction. Also, the words which occur sparsely like symbols and numbers have been ignored to get a uniform polarity distribution.
example : GVyH8kR9Dk

## Question 4 (5 points)

Even after filtering out words with count < 500, many top-most and bottom-most polarity don't make sense. Identify what kind of words these are and what can be done to filter them out. You can read some tweets in the file to see what's happening. 

# Part-2: Distributional Semantics

## Cosine Similarity

Recall that, where $i$ indexes over the context types, cosine similarity is defined as follows. $x$ and $y$ are both vectors of context counts (each for a different word), where $x_i$ is the count of context $i$.

$$cossim(x,y) = \frac{ \sum_i x_i y_i }{ \sqrt{\sum_i x_i^2} \sqrt{\sum_i y_i^2} }$$

The nice thing about cosine similarity is that it is normalized: no matter what the input vectors are, the output is between 0 and 1. One way to think of this is that cosine similarity is just, um, the cosine function, which has this property (for non-negative $x$ and $y$). Another way to think of it is, to work through the situations of maximum and minimum similarity between two context vectors, starting from the definition above.

Note: a good way to understand the cosine similarity function is that the numerator cares about whether the $x$ and $y$ vectors are correlated. If $x$ and $y$ tend to have high values for the same contexts, the numerator tends to be big. The denominator can be thought of as a normalization factor: if all the values of $x$ are really large, for example, dividing by the square root of their sum-of-squares prevents the whole thing from getting arbitrarily large. In fact, dividing by both these things (aka their norms) means the whole thing can’t go higher than 1.

In this problem we'll work with vectors of raw context counts.  (As you know, this is not necessarily the best representation.)

## Question 5 (5 points)

See the file `nytcounts.university_cat_dog`, which contains context count vectors for three words: “dog”, “cat”, and “university”. These are immediate left and right contexts from a New York Times corpus. You can open the file in a text editor since it’s quite small. 

Write a function which takes context count dictionaries of two words and calculates cosine similarity between these two words. The function should return a number beween 0 and 1.  Briefly comment on whether the relative simlarities make sense.



In [25]:
import distsim; reload(distsim)
import math

word_to_ccdict = distsim.load_contexts("nytcounts.university_cat_dog")

#word_to_ccdict['university']
def cosinesimilarity(word1,word2):
    set_common=set(word_to_ccdict[word1].keys()).intersection(word_to_ccdict[word2].keys())
    #print set_common
    numerator=0
    x_denom=0
    y_denom=0
    denominator=0
    for context in set_common:
        numerator+= word_to_ccdict[word1][context]*word_to_ccdict[word2][context]
    for w1cont in word_to_ccdict[word1].keys():
        x_denom+=word_to_ccdict[word1][w1cont]*word_to_ccdict[word1][w1cont]
    for w2cont in word_to_ccdict[word2].keys():
        y_denom+=word_to_ccdict[word2][w2cont]*word_to_ccdict[word2][w2cont]
    denominator=math.sqrt(x_denom)*math.sqrt(y_denom)    
    return numerator/denominator    
         
    
cosinesimilarity("cat","dog")    
    


file nytcounts.university_cat_dog has contexts for 3 words


0.9668916727145297


## Question 6 (20 points)

Explore similarities in `nytcounts.4k`, which contains context counts for about 4000 words in a sample of New York Times articles. The news data was lowercased and URLs were removed. The context counts are for the 2000 most common words in twitter, as well as the most common 2000 words in the New York Times. (But all context counts are from New York Times.) The context counts only contain contexts that appeared for more than one word.  The file has three tab-separate fields: the word, its count, and a JSON-encoded dictionary of its context counts.  You'll see it's just counts of the immediate left/right neighbors.

Choose **six** words. For each, show the output of 20 nearest words (use cosine similarity as distance metric). Comment on whether the output makes sense. Comment on whether this approach to distributional similarity makes more or less sense for certain terms.
Four of your words should be:

 * a name (for example: person, organization, or location)
 * a common noun
 * an adjective
 * a verb

You may also want to try exploring further words that are returned from a most-similar list from one of these. You can think of this as traversing the similarity graph among words.

*Implementation note:* 
On my laptop it takes several hundred MB of memory to load it into memory from the `load_contexts()` function. If you don’t have enough memory available, your computer will get very slow because the OS will start swapping. If you have to use a machine without that much memory available, you can instead implement in a streaming approach by using the `stream_contexts()` generator function to access the data; this lets you iterate through the data from disk, one vector at a time, without putting everything into memory. You can see its use in the loading function. (You could also alternatively use a key-value or other type of database, but that’s too much work for this assignment.)

In [26]:
import distsim; reload(distsim)
import operator
word_to_ccdict = distsim.load_contexts("nytcounts.4k")
#print word_to_ccdict.keys()
selected_words=["boston","school","shoot","fantastic","pizza","taylor"]
top_20={}

for word in selected_words:
    word_cosine={}
    for word_cont in word_to_ccdict.keys():
        if word_cont != word:
            word_cosine[word_cont]=cosinesimilarity(word,word_cont)
    top_20[word]= sorted(word_cosine.items(), key=operator.itemgetter(1),reverse=True)[:20]


for k, v in top_20.iteritems():
    print "\n", k
    c=0
    for i in v:
        c+=1
        print c,".",i 
    
        
    

#word_to_ccdict.keys()
###Provide your answer below; perhaps in another cell so you don't have to reload the data each time

file nytcounts.4k has contexts for 3648 words

shoot
1 . ('eat', 0.9631932558519338)
2 . ('kill', 0.9583701499962417)
3 . ('marry', 0.9557866418552138)
4 . ('sell', 0.9557345489364582)
5 . ('buy', 0.9550833034403917)
6 . ('win', 0.9547472671236099)
7 . ('stop', 0.9517573306977142)
8 . ('watch', 0.9510502330396143)
9 . ('steal', 0.9508007096647293)
10 . ('draw', 0.9492683767150271)
11 . ('treat', 0.948323109720575)
12 . ('produce', 0.9479117665578529)
13 . ('teach', 0.9465294538375284)
14 . ('write', 0.9463509205405134)
15 . ('hide', 0.9444941416115654)
16 . ('help', 0.9423991387357697)
17 . ('develop', 0.9418471624342122)
18 . ('pass', 0.9408048440585463)
19 . ('keep', 0.939019673271398)
20 . ('prevent', 0.9390169517589975)

school
1 . ('schools', 0.7410965056825047)
2 . ('college', 0.7161614959732483)
3 . ('line', 0.694353893026886)
4 . ('church', 0.692936169260658)
5 . ('practice', 0.6926395404877579)
6 . ('experience', 0.6896426171195871)
7 . ('location', 0.6868966350791714)
8 . ('s

## Question 7 (10 points)

In the next several questions, you'll examine similarities in trained word embeddings, instead of raw context counts.

See the file `nyt_word2vec.university_cat_dog`, which contains word embedding vectors pretrained by word2vec [1] for three words: “dog”, “cat”, and “university”, from the same corpus. You can open the file in a text editor since it’s quite small.

Write a function which takes word embedding vectors of two words and calculates cossine similarity between these 2 words. The function should return a number beween -1 and 1. Briefly comment on whether the relative simlarities make sense. 

*Implementation note:*
Notice that the inputs of this function are numpy arrays (v1 and v2). If you are not very familiar with the basic operation in numpy, you can find some examples in the basic operation section here:
https://docs.scipy.org/doc/numpy-dev/user/quickstart.html

If you know how to use Matlab but haven't tried numpy before, the following link should be helpful:
https://docs.scipy.org/doc/numpy-dev/user/numpy-for-matlab-users.html

[1] Mikolov, Tomas, et al. "Distributed representations of words and phrases and their compositionality." NIPS 2013.

In [35]:
import distsim; reload(distsim)
import numpy as np

word_to_vec_dict = distsim.load_word2vec("nyt_word2vec.4k")

def cosine_w2vec(word1,word2):
    v1=word_to_vec_dict[word1]
    v2=word_to_vec_dict[word2]
    num=v1.dot(v2)
    denom=np.sqrt(np.sum(np.square(v1)))*np.sqrt(np.sum(np.square(v2)))
    return num/denom
    
    #print v1,"\n",v2
cosine_w2vec("man","dog")
# write code here to show output (i.e. cosine similarity between these words.)
# We encourage you to write other functions in distsim.py itself. 

0.58011277913661741

The context score can ve generally predicted by how much one word can be substituted with another word and how much that sentence would make sense if that substitution is made.

cosine_w2vec("philadelphia","boston") -> 0.83843146669964441

High scores makes sense because of the similarities that both of them are places and Proper nouns.

cosine_w2vec("dog","boston") -> -0.094315197070464749

Both are not connected in any ways directly or remotely indirectly. Hence the low score.

cosine_w2vec("man","girl") -> 0.75017311159545375

Both have somewhat high scores since they both represent a person or human being

cosine_w2vec("man","dog") -> 0.58011277913661741

Both have compairitively low scores since one is a human and other is an animal

## Question 8 (20 points)

Repeat the process you did in the question 6, but now use dense vector from word2vec. Comment on whether the outputs makes sense. Compare the outputs of using nearest words on word2vec and the outputs on sparse context vector (so we suggest you to use the same words in question 6). Which method works better on the query words you choose. Please brief explain why one method works better than other in each case.

Not: we used the default parameters of word2vec in [gensim](https://radimrehurek.com/gensim/models/word2vec.html) to get word2vec word embeddings.

In [98]:
import distsim
word_to_vec_dict = distsim.load_word2vec("nyt_word2vec.4k")
selected_words=["boston","school","shoot","fantastic","pizza","taylor"]
for word in selected_words:
    word_cosine={}
    for word_cont in word_to_vec_dict.keys():
        if word_cont != word:
            word_cosine[word_cont]=cosine_w2vec(word,word_cont)
    top_20[word]= sorted(word_cosine.items(), key=operator.itemgetter(1),reverse=True)[:20]


for k, v in top_20.iteritems():
    print "\n", k
    c=0
    for i in v:
        c+=1
        print c,".",i 
    

###Provide your answer below


shoot
1 . ('throw', 0.83991503602688433)
2 . ('knock', 0.74445533245063233)
3 . ('steal', 0.74040588086964498)
4 . ('catch', 0.73487979786513447)
5 . ('fly', 0.72376486247027527)
6 . ('rip', 0.7216874028857152)
7 . ('pull', 0.72065890137358191)
8 . ('kick', 0.71650064755404108)
9 . ('beat', 0.69062492692197763)
10 . ('shake', 0.68865684831888441)
11 . ('hang', 0.68514175720810488)
12 . ('snap', 0.68325436939758821)
13 . ('walk', 0.67996269010607324)
14 . ('kill', 0.67858330357863184)
15 . ('burn', 0.67453166559025968)
16 . ('grab', 0.67315293863800429)
17 . ('shout', 0.67050749602171078)
18 . ('sit', 0.66752553861657993)
19 . ('stick', 0.65732212358984266)
20 . ('jump', 0.63513735318490838)

school
1 . ('schools', 0.75222831831959791)
2 . ('college', 0.74907314124814695)
3 . ('class', 0.62663670792003356)
4 . ('student', 0.5800215510507376)
5 . ('classes', 0.55825512394961097)
6 . ('columbia', 0.54627428159697367)
7 . ('teacher', 0.53680636671823401)
8 . ('academy', 0.5340936830273453

## Question 9 (15 points)
An interesting thing to try with word embeddings is analogical reasoning tasks. In the following example, it's intended to solve the analogy question "king is to man as what is to woman?", or in SAT-style notation,

king : man :: ____ : woman

Some research has proposed to use additive operations on word embeddings to solve the analogy: take the vector $(v_{king}-v_{man}+v_{woman})$ and find the most-similar word to it.  One way to explain this idea: if you take "king", get rid of its attributes/contexts it shares with "man", and add in the attributes/contexts of "woman", hopefully you'll get to a point in the space that has king-like attributes but the "man" ones replaced with "woman" ones.

Show the output for 20 closest words you get by trying to solve that analogy with this method.  Did it work?

Please come up with another analogical reasoning task (another triple of words), and output the answer using the same method. Comment on whether the output makes sense. If the output makes sense, explain why we can capture such relation between words using an unsupervised algorithm. Where does the information come from? On the other hand, if the output does not make sense, propose an explanation why the algorithm fails on this case.


Note that the word2vec is trained in an unsupervised manner just with distributional statistics; it is interesting that it can apparently do any reasoning at all.  For a critical view, see [Linzen 2016](http://www.aclweb.org/anthology/W/W16/W16-2503.pdf).



In [37]:
import distsim
word_to_vec_dict = distsim.load_word2vec("nyt_word2vec.4k")
#print word_to_vec_dict.keys()
analogy=["king","man","woman"]
analogy_vector=word_to_vec_dict[analogy[0]]- word_to_vec_dict[analogy[1]] + word_to_vec_dict[analogy[2]]
word_cosine={}

top_20={}
def cosine_w2vecSAT(word1,word2):
    v2=word_to_vec_dict[word2]
    num=word1.dot(v2)
    denom=np.sqrt(np.sum(np.square(word1)))*np.sqrt(np.sum(np.square(v2)))
    return num/denom


for word_cont in word_to_vec_dict.keys():
        if word_cont not in analogy:
            word_cosine[word_cont]=cosine_w2vecSAT(analogy_vector,word_cont)
top_20= sorted(word_cosine.items(), key=operator.itemgetter(1),reverse=True)[:20]        
    
count=0
for i in top_20:
    count+=1
    print count,".",i





1 . ('queen', 0.72502863198579404)
2 . ('princess', 0.57790010340102305)
3 . ('prince', 0.56696239241698621)
4 . ('lord', 0.53091939111112818)
5 . ('royal', 0.52020329686379774)
6 . ('mary', 0.49769814628352954)
7 . ('mama', 0.49546963683192263)
8 . ('daughter', 0.49375794656560135)
9 . ('singer', 0.4898380820139282)
10 . ('kim', 0.48835469524343839)
11 . ('elizabeth', 0.48248484340535652)
12 . ('girl', 0.47733829480775547)
13 . ('grandma', 0.47699072668127201)
14 . ('sister', 0.47030437182473767)
15 . ('mother', 0.4694220288325171)
16 . ('clark', 0.46824004740972941)
17 . ('wedding', 0.46233629356038564)
18 . ('husband', 0.45685118817885451)
19 . ('boyfriend', 0.44755057450353947)
20 . ('jesus', 0.43857211580612976)
